cluster-autoscaler/processors/nodeinfosprovider/mixed_nodeinfos_processor.go

```diff
@@ -95,8 +95,15 @@ func (p *MixedTemplateNodeInfoProvider) Process(ctx *context.AutoscalingContext,
 		id := nodeGroup.Id()
 		if _, found := result[id]; !found {
 			// Build nodeInfo.
-			nodeInfo, err := simulator.BuildNodeInfoForNode(node, podsForNodes[node.Name], daemonsets, p.forceDaemonSets)
-			sanitizedNodeInfo, err := utils.SanitizeNodeInfo(nodeInfo, id, ignoredTaints)
+			sanitizedNode, err := utils.SanitizeNode(node, id, ignoredTaints, unwantedTaints)
+			if err != nil {
+				return false, "", err
+			}
+			nodeInfo, err := simulator.BuildNodeInfoForNode(sanitizedNode, podsForNodes[node.Name], daemonsets, p.forceDaemonSets)
+			if err != nil {
+				return false, "", err
+			}
+			sanitizedNodeInfo, err := utils.SanitizeNodeInfo(nodeInfo, id, ignoredTaints, unwantedTaints)
```

Comment:
This isn't really a part of your changes, but with the extra SanitizeNode call it's just more visible - why is the above logic different from the GetNodeInfoFromTemplate()? Doesn't it do the exact same thing just with different nodeInfo as a source?

I understand this may not be in the scope of this PR, so feel free not to do it, but maybe it would be worth a small refactor to unify those codepaths?

Reply:
That's a valid point, but this is not 100% true. We treat DS pods slightly differently depending on where we get template from. For existing nodes we take DS pods scheduled them and fill with missing DS pods if the option `--force-ds` is enabled. For instance template ones we always pack all the DS pods.

I think long term solution here is to set `--force-ds` to true always and unify both these logic pieces. But for now I'm not sure if it won't obfuscate the different underlying behaviour.